src/python/pants/backend/scala/resolve/lockfile.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from pants.backend.scala.dependency_inference.symbol_mapper import AllScalaTargets
+from pants.backend.scala.subsystems.scala import ScalaSubsystem
+from pants.engine.rules import collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.jvm.goals.lockfile import (
+    ValidatedJvmArtifactsForResolve,
+    ValidateJvmArtifactsForResolveRequest,
+)
+from pants.jvm.resolve.common import Coordinate
+from pants.jvm.subsystems import JvmSubsystem
+
+SCALA_LIBRARY_GROUP = "org.scala-lang"
+SCALA_LIBRARY_ARTIFACT = "scala-library"
+
+
+class ConflictingScalaLibraryVersionInResolveError(ValueError):
+    """Exception for when there is a conflicting Scala version in a resolve."""
+
+    def __init__(
+        self, resolve_name: str, required_version: str, conflicting_coordinate: Coordinate
+    ) -> None:
+        super().__init__(
+            f"The JVM resolve `{resolve_name}` contains a `jvm_artifact` for version {conflicting_coordinate.version} "
+            f"of the Scala runtime. This conflicts with Scala version {required_version} which is the configured version "
+            "of Scala for this resolve from the `[scala].version_for_resolve` option. "
+            "Please remove the `jvm_artifact` target with JVM coordinate "
+            f"{conflicting_coordinate.to_coord_str()}, then re-run the `generate-lockfiles` goal."
+        )
+
+
+class MissingScalaLibraryInResolveError(ValueError):
+    def __init__(self, resolve_name: str, scala_version: str) -> None:
+        super().__init__(
+            f"The JVM resolve `{resolve_name}` does not contain a requirement for the Scala runtime. "
+            "Since at least one Scala target type in this repository consumes this resolve, the resolve "
+            "must contain a `jvm_artifact` target for the Scala runtime.\n\n"
+            "Please add the following `jvm_artifact` target somewhere in the repository and re-run "
+            "the `generate-lockfiles` goal:\n"
+            "jvm_artifact(\n"
+            f'  name="{SCALA_LIBRARY_GROUP}_{SCALA_LIBRARY_ARTIFACT}_{scala_version}",\n'
+            f'  group="{SCALA_LIBRARY_GROUP}",\n',
+            f'  artifact="{SCALA_LIBRARY_ARTIFACT}",\n',
+            f'  version="{scala_version}",\n',
+            f'  resolve="{resolve_name}",\n',
+            ")",
+        )
+
+
+class ValidateResolveHasScalaRuntimeRequest(ValidateJvmArtifactsForResolveRequest):
```

Comment:
Ditto on `Validate` vs `Validated`. You could use `ValidateResult`

Reply:
Renamed the output type to `ValidateJvmArtifactsForResolveResult`.